syz-verifier/verifier.go

```diff
@@ -49,6 +52,166 @@ type Verifier struct {
 	statsWrite    io.Writer
 	newEnv        bool
 	reruns        int
+
+	// We use single queue for every Kernel environment.
+	KernelEnvTasks [][]chan *ExecTask
+}
+
+func (vrf *Verifier) Init() {
+	vrf.KernelEnvTasks = make([][]chan *ExecTask, len(vrf.pools))
+	for i := range vrf.KernelEnvTasks {
+		vrf.KernelEnvTasks[i] = make([]chan *ExecTask, EnvironmentsCount)
+	}
+
+	for i := range vrf.KernelEnvTasks {
+		for j := range vrf.KernelEnvTasks[i] {
+			vrf.KernelEnvTasks[i][j] = make(chan *ExecTask, 100)
+		}
+	}
+
+	srv, err := startRPCServer(vrf)
+	if err != nil {
+		log.Fatalf("failed to initialise RPC server: %v", err)
+	}
+	vrf.srv = srv
+}
+
+func (vrf *Verifier) ProgramsWaitingForRunners() (res int) {
+	for i := range vrf.KernelEnvTasks {
+		for j := range vrf.KernelEnvTasks[i] {
+			res += len(vrf.KernelEnvTasks[i][j])
+		}
+	}
+	return
+}
+
+func (vrf *Verifier) StartProgramsAnalysis() {
+	go func() {
+		vrf.srv.cond.L.Lock()
+		vrf.srv.cond.Wait()
+		vrf.srv.cond.L.Unlock()
+
+		type AnalysisResult struct {
+			Diff []*ExecResult
+			Prog *prog.Prog
+		}
+		results := make(chan *AnalysisResult)
+
+		for {
+			for vrf.ProgramsWaitingForRunners() < 100 {
+				go func(prog *prog.Prog) {
+					results <- &AnalysisResult{
+						vrf.TestProgram(prog),
+						prog,
+					}
+				}(vrf.generate())
+			}
+
+			select {
+			case result := <-results:
+				if result.Diff != nil {
+					vrf.SaveDiffResults(result.Diff, result.Prog)
+				}
+			case <-time.After(1 * time.Second):
+			}
+		}
+	}()
+}
+
+func (vrf *Verifier) GetRunnerTask(kernel int, existing EnvDescr) *rpctype.RPCExecTask {
+	for env := existing; env >= AnyEnvironment; env-- {
+		select {
+		case task, ok := <-vrf.KernelEnvTasks[kernel][env]:
```

Comment:
Receive returns false in the second argument iff the channel is closed, we never close it. So just:
```
		case task := <-vrf.KernelEnvTasks[kernel][env]:
			return task.ToRPC()
```

Reply:
Hm... Thanks, I'll read the documentation here.

Reply:
Deleted.